File: Core-DSA-Optimized/Strings/Z-Algorithm/z_algorithm.py
class ZAlgorithm:
    @staticmethod
    def _get_z_array(s: str) -> list[int]:
        n = len(s)
        z = [0] * n
        # [L, R] represents the current bounding box of the rightmost prefix match
        L, R = 0, 0
        
        for i in range(1, n):
            # If i is within the bounding box, we can copy the previously calculated value
            if i <= R:
                z[i] = min(R - i + 1, z[i - L])
                
            # Manually expand the box if necessary
            while i + z[i] < n and s[z[i]] == s[i + z[i]]:
                z[i] += 1
                
            # If the new match extends past our current right bound, update the box
            if i + z[i] - 1 > R:
                L = i
                R = i + z[i] - 1
                
        return z

    @staticmethod
    def search(pattern: str, text: str) -> list[int]:
        matches = []
        if not pattern or not text:
            return matches

        # Concatenate pattern and text with a unique separator
        # Note: In a true ultra-low-memory environment, we would virtualize this!
        concat = pattern + "$" + text
        m = len(pattern)
        
        z_array = ZAlgorithm._get_z_array(concat)
        
        # Any Z-value exactly equal to the pattern length is a match
        for i in range(len(z_array)):
            if z_array[i] == m:
                # Calculate original index in the text: i - length(pattern) - length("$")
                matches.append(i - m - 1)
                
        return matches

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    text = "AABAACAADAABAABA"
    pattern = "AABA"
    
    results = ZAlgorithm.search(pattern, text)
    
    # "AABA" appears at indices 0, 9, and 12
    assert results == [0, 9, 12], f"Failed! Expected [0, 9, 12], got {results}"
    
    print("Python Z-Algorithm Pattern Matching Test Passed!")

File: Core-DSA-Optimized/Strings/Z-Algorithm/z_algorithm.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>

using namespace std;

class ZAlgorithm {
private:
    static vector<int> getZArray(const string& s) {
        int n = s.length();
        vector<int> z(n, 0);
        int L = 0, R = 0;

        for (int i = 1; i < n; i++) {
            if (i <= R) {
                z[i] = min(R - i + 1, z[i - L]);
            }

            while (i + z[i] < n && s[z[i]] == s[i + z[i]]) {
                z[i]++;
            }

            if (i + z[i] - 1 > R) {
                L = i;
                R = i + z[i] - 1;
            }
        }

        return z;
    }

public:
    static vector<int> search(const string& pattern, const string& text) {
        vector<int> matches;
        if (pattern.empty() || text.empty()) return matches;

        string concat = pattern + "$" + text;
        int m = pattern.length();
        vector<int> zArray = getZArray(concat);

        for (int i = 0; i < zArray.size(); i++) {
            if (zArray[i] == m) {
                matches.push_back(i - m - 1);
            }
        }

        return matches;
    }
};

// --- CI/CD Automated Test ---
int main() {
    string text = "AABAACAADAABAABA";
    string pattern = "AABA";

    vector<int> results = ZAlgorithm::search(pattern, text);

    assert(results.size() == 3);
    assert(results[0] == 0);
    assert(results[1] == 9);
    assert(results[2] == 12);

    cout << "C++ Z-Algorithm Pattern Matching Test Passed!\n";
    return 0;
}